src/Databases/DDLDependencyVisitor.cpp

```diff
@@ -1,198 +1,239 @@
 #include <Databases/DDLDependencyVisitor.h>
 #include <Dictionaries/getDictionaryConfigurationFromAST.h>
 #include <Interpreters/Context.h>
-#include <Interpreters/evaluateConstantExpression.h>
 #include <Parsers/ASTCreateQuery.h>
 #include <Parsers/ASTFunction.h>
 #include <Parsers/ASTIdentifier.h>
 #include <Parsers/ASTLiteral.h>
 #include <Parsers/ASTSelectWithUnionQuery.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
 #include <Poco/String.h>
 
-namespace DB
-{
 
-using TableLoadingDependenciesVisitor = DDLDependencyVisitor::Visitor;
-
-TableNamesSet getDependenciesSetFromCreateQuery(ContextPtr global_context, const QualifiedTableName & table, const ASTPtr & ast)
-{
-    assert(global_context == global_context->getGlobalContext());
-    TableLoadingDependenciesVisitor::Data data;
-    data.default_database = global_context->getCurrentDatabase();
-    data.create_query = ast;
-    data.global_context = global_context;
-    TableLoadingDependenciesVisitor visitor{data};
-    visitor.visit(ast);
-    data.dependencies.erase(table);
-    return data.dependencies;
-}
-
-void DDLDependencyVisitor::visit(const ASTPtr & ast, Data & data)
+namespace DB
 {
-    /// Looking for functions in column default expressions and dictionary source definition
-    if (const auto * function = ast->as<ASTFunction>())
-        visit(*function, data);
-    else if (const auto * dict_source = ast->as<ASTFunctionWithKeyValueArguments>())
-        visit(*dict_source, data);
-    else if (const auto * storage = ast->as<ASTStorage>())
-        visit(*storage, data);
-}
 
-bool DDLMatcherBase::needChildVisit(const ASTPtr & node, const ASTPtr & child)
+namespace
 {
-    if (node->as<ASTStorage>())
-        return false;
-
-    if (auto * create = node->as<ASTCreateQuery>())
+    /// CREATE TABLE or CREATE DICTIONARY or CREATE VIEW or CREATE TEMPORARY TABLE or CREATE DATABASE query.
+    void visitCreateQuery(const ASTCreateQuery & create, DDLDependencyVisitor::Data & data)
     {
-        if (child.get() == create->select)
-            return false;
-    }
+        QualifiedTableName to_table{create.to_table_id.database_name, create.to_table_id.table_name};
+        if (!to_table.table.empty())
+        {
+            /// TO target_table (for materialized views)
+            if (to_table.database.empty())
+                to_table.database = data.default_database;
```

Comment:
It should not be empty, because the query [is normalized ](https://github.com/ClickHouse/ClickHouse/blob/b065f9747dd39cca61cd6573a791a49e4b89220e/src/Interpreters/InterpreterCreateQuery.cpp#L1146-L1147). Btw, can `default_database` be different from `current_database`? (Seems like it can)

Reply:
It must be the same for the global database. I just thought maybe it's better to check dependencies for a backup in the query context, not in the global context, so probably I'll replace that with the current database.

Reply:
I changed the code so it now uses the current database - https://github.com/ClickHouse/ClickHouse/pull/44158